### PR TITLE
Add support for hash-param

### LINKED ISSUE #13: jump to queue index/item, hash-param for queue index
When navigating the queue and refreshing, the user gets sent back to queue index 1/X regardless of position before refresh.
It would be preferable if we could somehow restore the previous position in the queue.

However, just storing the index of the queue in the hash-param might not always be right, since the list itself will be refreshed and might have changed. (Item was previously at index 10, previous items got removed, item is now at index 7 but user ends up at index 10)

Also, if the user just acted on an item in the queue, which caused it to be removed from the queue, we should end up on the same index but another item. (Item was previously at index 10, item got removed, user ends up on index 10 - next item)

Both combined would be item at index 10, previous items got removed, item should have been at index 7 but got removed as well so user should end up at index 7.

Not sure how to solve this best.
One idea is to save the item id in hash-param and look for it when items change.
Although, since we don't have separate idList and items anymore, we lose the id when we get the item model.
However, to get to the index the item _would have had_ we need to save an adjacent (next?) id as a fallback id to jump to in case the primary id is missing.

We would also use this "jump" functionality in list/queue views where the user wants to jump to a specific item.

The base of it all might be to implement:

- [x] `selectIndex(queueIndex)`
- [ ] `selectIndexByItem(item)`
- [ ] `selectIndexById(id)`

### LINKED ISSUE #41: Add option to keep selected item selected when data changes
Add option to keep selected item selected when data changes.

### PR DESCRIPTION
**Pull request by megheaiulian**
fixes #13
fixes #41 when `maintainSelection` is `true`

### CODE REVIEW

**Comment by megheaiulian** [comment]
@nomego This is ready for review.

**Comment by nomego** [comment]
I assume this needs a rebase now ?

**Comment by megheaiulian** [comment]
This doesn't need a rebase. Tests PR will need a rebase but after this is merged.

File: cosmoz-data-nav.js [hunk 1]
```diff
@@ -63,6 +63,12 @@
 				readOnly: true
 			},
 
+			_hasItems: {
```
Comment (nomego) [inline cosmoz-data-nav.js]: Why prefixed with _ ? Seems shaky for CSS?

File: .codeclimate.yml [hunk 2]
```diff
@@ -1,28 +1,49 @@
 ---
-engines:
-  eslint:
+version: "2"
```
Comment (nomego) [inline .codeclimate.yml]: Please also do a PR for the docs for how we configure the code climate parameters with refs to docs or explanations to the options

**Review by nomego — APPROVED** [review]

**Review by plequang — APPROVED** [review]
Looks good.
Tested various use cases in real app and seems to work nicely.
Though I quite agree with @nomego regarding `_hasItems`.
What do you say @megheaiulian ?

**Comment by megheaiulian** [comment]
@plequang @nomego 
_hasItems is used in css to show only the first element and it's incomplete content when `items` length is 0.
In CSS there is something like this:
```css
  :host(:not([_has-items])) #items > ::slotted(:not(:first-of-type)){
     display: none;
   }
```
That basically shows the first element only when there is no data.
When it is doable like that I think it's preferable over doing in code as we can change easily only the CSS.

Let me know what you think.

**Comment by plequang** [comment]
@megheaiulian Yes sure, and that's  correct.
But I think @nomego refers to the attribute name starting with an underscore `_` and used in the css selector.
Long time ago (yes, @nomego  is old), there used to be some  Internet Explorer hacks with names starting with an underscore if I remember well. Maybe that's why he thinks it's "shaky".

**Comment by plequang** [comment]
So what do we do with that PR ?
Should `_hasItems` be renamed ?
@nomego @megheaiulian

**Comment by megheaiulian** [comment]
@nomego I've updated the PR. Can you have a look at the changes?

File: cosmoz-data-nav.js [hunk 3]
```diff
@@ -172,6 +187,30 @@
...
+			 */
+			hashParam: {
+				type: String
+			},
+
+			/**
+			 * The route hash parameters extracted by the `cosmoz-page-location`
+			 * element.
+			 */
+			_routeHashParams: {
```
Comment (nomego) [inline cosmoz-data-nav.js:203]: Why notify on private property?

File: cosmoz-data-nav.js [hunk 4]
```diff
@@ -411,10 +443,19 @@
 				});
 			}
 
-			if (this.selected === 0) {
+			if (this._updateSelectedFromHash()) {
+				return;
+			}
+
+			const index = this.maintainSelection && this.selected > 0 ? Math.max(0, items
```
Comment (nomego) [inline cosmoz-data-nav.js]: Set the `Math.max()` result to a const of its own for readability.

File: cosmoz-data-nav.js [hunk 5]
```diff
@@ -516,8 +558,8 @@
 			};
 		},
 
-		_getElement(index, elements = this._elements) {
-			return elements[index % this.elementsBuffer];
+		_getElement(index) {
```
Comment (nomego) [inline cosmoz-data-nav.js]: Re-add the elements arg?

File: cosmoz-data-nav.js [hunk 6]
```diff
@@ -859,6 +887,54 @@
...
+
+		_getItemId(item) {
+			return this.isIncompleteFn(item) ? item  : this.get(this.idPath, item);
+		},
+
+		_updateHashForSelected(selected) {
+			const hashParam = this.hashParam,
+				idPath = this.idPath;
+
+			if (!(hashParam && idPath && this._routeHashParams && this.items.length)) {
```
Comment (nomego) [inline cosmoz-data-nav.js]: I'd like the revert approach, `!hashParam || !idPath .. `

**Review by nomego — APPROVED** [review]

File: cosmoz-data-nav.js (hunk 3 above)
Comment (megheaiulian) [reply cosmoz-data-nav.js:203]: Removed notify

**Pull request merged**